### ISSUE TITLE
Random: Unable to list images in local Docker cache and connection refused to Nomad on remote builder runner

### ISSUE THREAD

**Issue by michiels** [hashicorp/waypoint#4189]
**Describe the bug**
I'm trying to run `waypoint up` or `waypoint build` with Nomad as platform and a remote waypoint runner installed. Also using HCP Waypoint Beta as Waypoint server.

Whenever I run `waypoint up` or `waypoint build`, I get either of the following three results:

```
$ waypoint build 
» Building firmhouse...
! Put "http://localhost:4646/v1/jobs?namespace=default&region=global": dial tcp
  127.0.0.1:4646: connect: connection refused
```

```
$ waypoint build
» Building firmhouse...
! unable to list images in local Docker cache: Cannot connect to the Docker daemon
  at unix:///var/run/docker.sock. Is the docker daemon running?
```

Sometimes it seems I don't get the Nomad connection error or Docker cache error and the command continues, but then it seems to wait for a runner assignment indefinitely:

```
$ waypoint build
» Building firmhouse...

» Operation is queued waiting for job "01GHGDD8KK4EBX73JZKE99Z7PX". Waiting for runner assignment...
  If you interrupt this command, the job will still run in the background.
```

I can see in the runner build logs in the Nomad cluster. Some things are happening and it seems the above errors come from the runner.

**Steps to Reproduce**
Steps to reproduce the behavior.

Gist with `waypoint.hcl` and output logs from `waypoint build` and remote runner: https://gist.github.com/michiels/92e80b7dce294ec512eae8f5bd674810

**Expected behavior**
Waypoint to run build and deploy steps via remote runner on Nomad cluster.

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: v0.10.3 (136a885b9)
* Waypoint Server Platform and Version: Waypoint server on HCP. Nomad v1.4.2 on cluster.
* Waypoint Plugin: `docker-pull` and `nomad`

**Comment by michiels** [comment]
I realize now that (part of) some of these errors is that HCP Waypoint doesn't know how to talk to my Nomad cluster. I wonder if there is a way to update/tell the HCP Waypoint server/project which Nomad Address and authentication would have to be used.

It doesn't completely explain however the error about Docker not being found. Or why some jobs seem to hang in some kind of pending state.

**Comment by briancain** [comment]
Hey there @michiels - So HCP Waypoint server shouldn't be talking to your personal Nomad cluster. I believe the runner you install is the actual piece of software that talks to your Nomad cluster. How did you install your Waypoint runner, and does it have access to Nomad?

**Comment by michiels** [comment]
Hi @briancain. Thanks for the response!

This is the command I used to install the runner:

```
waypoint runner install \                                                                                              
  -platform=nomad \
  -server-addr=api.hashicorp.cloud:443 \
  -nomad-runner-image=hashicorp/waypoint \
  -nomad-host-volume=wp-runner-vol -id=default -project=firmhouse-hcp -app=firmhouse
```

I just did a complete new fresh setup with only 1 Nomad node (self-bootstrapped) and a new project on HCP. Still the same result:

From `waypoint build -local=false` runner log and I also see the same error on the `stderr` log on the Nomad node's job:

```
2022-11-13T09:45:53.418+0100 [WARN]  waypoint: job failed: job_id=01GHR2N71YBMYAVJGYVVASC9X4 code=FailedPrecondition message="unable to list images in local Docker cache: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
```

**Comment by michiels** [comment]
@briancain I think I'm a small step close. I turns out I still had like 15+ runner profiles on the server from older configurations, including some Docker-platform profiles. So what I think/assume happened  is that the runner randomly selected some of the runner profiles already on the server, and tried to connect to a local Docker server.

I've now removed all runner profiles and created a single new one, targeting my now only runner:

```
waypoint runner profile set -target-runner-id=default -plugin-type=nomad -default -project=firmhouse-hcp -app=firmhouse
```

Now I seem to get a consistent error, which is that the runner can't find the local nomad on port `4646`:

```2022-11-13T10:34:14.258+0100 [WARN]  waypoint: job failed: job_id=01GHR5DQY2ABYSJVJ3P2DD1PSE code=Unknown message="Put "http://localhost:4646/v1/jobs?namespace=default&region=global": dial tcp 127.0.0.1:4646: connect: connection refused"

! Put "http://localhost:4646/v1/jobs?namespace=default&region=global": dial tcp
  127.0.0.1:4646: connect: connection refused
```

This would make sense I guess because the nomad server would not be accessed via `localhost:4646` from within the Runner container. But I'm a bit too unfamiliar with Docker networking and the waypoint runner software to know if this is the right assumption.

I tried updating the `NOMAD_ADDR` on the runner by setting that ENV var in the profile, but that seems to not be having effect:

```
waypoint runner profile set -target-runner-id=default -plugin-type=nomad -default -project=firmhouse-hcp -app=firmhouse -env-var=NOMAD_ADDR=http://public-nomad-host:4646/
```

(I don't have TLS or ACL set up yet so redacting the hostname).

The runner keeps trying to connect with `localhost:4646`. 

Is there any way to make the nomad port accessible from within the container or configure the nomad address it should connect to?

**Comment by michiels** [comment]
Ok update here. And at this point I'm using this GitHub issue as a rubber duck just so others can also follow along. And turn it into a blog post/guide ;-)

I figured out I can set the `NOMAD_ADDR` in runners via:

```
$ waypoint config set -scope=global -runner NOMAD_ADDR='http://public-nomad-host:4646/'
```

Now everything in terms of connection works. I'm still getting an error, but that's because the nodes in my cluster don't have the required free memory to run the build+deploy tasks. 

So going to do add a VM with larger memory to the Nomad cluster and will then report back if everything worked.

**Comment by briancain** [comment]
Hey @michiels - Thanks for filling in all of the debug details for your process! Let us know if adding the memory works and if that resolves the issue for you! If so, please feel free to close this issue.

**Comment by nullc4t** [comment]
@michiels Thank you so much for your research, you solution works perfectly!

**Comment by nullc4t** [comment]
@briancain There's definitely lack of documentation and/or tutorials about this

**Comment by briancain** [comment]
Hey @nullc4t ! If there's something specific in the docs or a tutorial you wish to see, please open an issue and we can take a look at tackling it! We are looking for ways to improve our docs so we appreciate any specific feedback you may have. Thank you!

**Comment by arbourd** [comment]
I don't think this can be resolved because this simply doesn't work out-of-the-box with the instructions provided by HCP Waypoint Beta: 
<img width="620" alt="Screenshot 2023-05-30 at 10 16 22 AM" src="https://github.com/hashicorp/waypoint/assets/7211830/36486162-9b66-40d0-a423-ef39f371fe55">.

Either the documentation for HCP Waypoint should be updated to reflect how to get the Waypoint container to talk to the host properly, or most likely, the `waypoint runner install` command should probably be improved to ensure that installing Waypoint on Nomad works.

Anything I'm missing?
